utility/service/README.md

```diff
@@ -0,0 +1,159 @@
+4) Ensure session block height is current
+5) Get the `sessionContext` to observe the world state at that version
+6) Get the application object from the `r.AAT()` (using `sessionContext`)
+7) Get session node count from that session height (using `sessionContext`)
+8) Get maximum possible relays for the application (using `sessionContext`)
+9) Ensure not over serviced (if max relays is exceeded, not compensated for further work)
+10) Generate the session from seed data (see Session Protocol)
+11) Validate self against the session (is node within session)
+```mermaid
+graph TD
+    A[Relay.Validate] -->B
+    B[HasValidPayload] -->|Yes| C
+    B -->|No| Z
+    C[HasValidMeta] -->|Yes| D
+    C -->|No| Z
+    D[RelayChainIsSupported]-->|Yes| E
+    D -->|No| Z
+    E[IsValidSessionHeight]-->|Yes| F
+    E -->|No| Z
+    F[IsAppOverServiced]-->|No| G
+    F -->|Yes| Z
+    G[IsValidSession]-->|Yes| H
+    H -->|No| Z
+    Z[Reject Invalid Relay]
+```
+
+### Store the `Relay`
+
+Store a submitted `Relay` by a client for volume tracking
+
+1) Marshal `Relay` object into protoBytes
+2) Calculate the `hashOf(protoBytes)` <needed for volume tracking>
+3) Persist `Relay` object, indexing under session
+
+```mermaid
+graph TD
+    A[Relay.Store] -->|Encode `Relay` object|B
+    B[RelayProtoBytes] -->|Calculate Hash for Volume Tracking| C
+    C[RelayProtoBytes.Hash] -->|Add Hash to `Relay` and Persist| D
+    D[Relay.Persist] --> |Indexing Under Session|E
+    E[Key:SessionKey Val: Relays.AddNew]
+```
+
+### Execute the `Relay`
+
+Execute a submitted `Relay` against the `RelayChain` by a client after validation
+
+1) Retrieve the `RelayChain` url from chains.json file
+2) Execute http request with the `Relay Payload`
+3) Format and digitally sign the response
+4) Send back to client
+
+##### Wait for Session to end / secret key to be revealed
+
+It's important to note, the secret key isn't revealed by the network until the session is over
+to prevent volume based bias. The secret key is usually a pseudorandom selection using the block hash as a seed. 
+(See the session protocol)
+
+
+### Get volume metric applicable `Relays` from store
+
+1) Pull all `Relays` whose hash collides with the revealed secret key
+
+`SELECT * from RELAY where HashOf(relay) ends with hashEndWith AND sessionBlockHeight=sessionBlockHeight`
+
+2) This function also signifies deleting the non-volume-applicable `Relays`
+
+### Report volume metric applicable relays to `Fisherman`
+
+1) All volume applicable relays need to be sent to the assigned trusted `Fisherman` for a proper verification of the volume completed.
+2) Send `volumeRelays` to `fishermanServiceURL` through http.
+
+```mermaid
+graph TD
+    A[Relay.Execute] -->|Lookup Relay.RelayChain URL|B
+    B[Chains.JSON -> RelayChainURL] -->|Execute Http Request| C
+    C[RelayChain] -->|Return Response| D
+    D[RelayResponse.Sign] --> |Send|E
+    E[Requester]
+```
+
+##### Alt Design
```

Comment:
Not a blocker and might be worth an offline discussion but just wanted to share an idea and am curious if:
1. This is something you're familiar with (or explored before) and know it's not going to work
2. You see any issues with the bat.

I've started looking into how Chia Network uses VDF (Verifiable Delay Functions) to have proof of space & time consensus (proof of committing storage to the network over some time). This work is based on Brem Cohen's Proof of Sequential work: https://eprint.iacr.org/2018/183.pdf.

In short, I'm not 100% if the math enables this yet, but if it does, this might be an alternative design as well, which could theoretically be simpler and more memory efficiency:

![mermaid-diagram-2022-10-22-204928](https://user-images.githubusercontent.com/1892194/197373225-983896d1-4ea8-426b-aeb7-752e733d3fed.png)

Reply:
Sure - always open to new ideas. Chia's Proof of Time and Space was researched a bit during the consensus V1 spec'ing phase.

Would be interested to hear what you have in mind

Reply:
The underlying piece of proof of time & space are Verifiable Delay Functions (VDFs) which are based on Proof of Sequential work proofs (which are not functions).

After doing some more research and digging, I came to a similar solution to what you had with probabilistic hash collisions, but using Solana's proof of history[1]:

![Screen Shot 2022-10-24 at 5 16 08 PM](https://user-images.githubusercontent.com/1892194/197655334-9a6b95aa-ca8a-4014-bec1-e7fcb8c2178f.png)

I realize this diagram is getting a bit complicated so we wouldn't commit it as such, but hopefully we can use it for a discussion between ourselves:

```mermaid
sequenceDiagram
	    autonumber
        title Relay Volume Counter
	    actor C as Client
	    actor SN as Service Node
        participant IS as Internal Storage (Service Node)
        actor F as Fisherman
        C->>C: Generate SessionKey<br>(i.e. pseudo-random seed)
        C->>C: Initialize N=0
        C->>SN: SessionKey
	    loop Repeat N times - once per relay during session duration
            C->>+SN: SignedRelay
            SN->>SN: Validate & Process Relay
            SN->>SN: countHash=hashFunc(append(relayHash, countHash-1))<br>(using VDF)
            alt every M relays (e.g. 1000)
                SN->>IS: store(countHash)
            end
            SN->>-C: (countHash, SignedRelayResponse)
            C->>C: N += 1<br>update countHash
	    end
        C->>F: (SessionKey, countHash, N)
        SN->>F: (SessionKey, [countHashes], M)
        F->>F: assert(M == N)<br> Verify(SessionKey, [countHashes])
```

In some ways, Proof of history was built for verifiably proving increasing counters which gives me a gut sense that it may be a pointer in the right direction. 

To make the verification more optimal, it requires fishermen to be equipped with GPUs (see section 4.3 in [1]).

In relation to the hash collision approach, I think this ultimately just gives us a more deterministic workflow but I'm no longer sure it buys us any other benefits. 

[1] https://solana.com/solana-whitepaper.pdf

**Out of scope:** Eventually (not on v1 launch), we could also leverage the VDF to have a higher difficulty for RPC calls that "should cost more" (like eth archival calls), but I think that's outside the scope of this discussion but just a thought I had.

Reply:
If we do go with VDFs, we should probably use a golang wrapper around Rust libraries. 

Luckily, harmony already paid someone to do it: https://gitcoin.co/issue/25313

Reply:
AFAIK this is a sub-optimal solution because it requires you to sequentially verify the hashes - am I misunderstanding?

Reply:
Will add a screenshot to the point I made earlier for more clarification.

> To make the verification more optimal, it requires fishermen to be equipped with GPUs (see section 4.3 in [1]).

<img width="564" alt="Screen Shot 2022-10-25 at 7 31 55 AM" src="https://user-images.githubusercontent.com/1892194/197802227-33982c42-88f5-467e-ad43-88e601a08d46.png">

This will allow verification through parallelized hash verification which is guaranteed to have been executed sequentially.

The improvements I see are:
1. Deterministic vs probabilistic
2. Only involves verifying that the counts from the ServiceNode and Client are the same followed by the hash verification with the session context as the seed. Alternative, in order to avoid an attack where the ServiceNode starts pre-computing lots of hashes during the duration of a session, we need to open up the relay, verify the hash collision, check the signature of the client on each relay, and have some sort of deduping mechanism in place as well (which is avoidable by the sequential nature of the VDF).

Reply:
Don't want to block it, but do think it's worth tracking the discussion and decision.

I can move it to https://github.com/pokt-network/pocket/discussions if need be

Reply:
https://github.com/pokt-network/pocket/discussions/330